### PR TITLE
Fix enabling / disabling of APM agent "recording" in APMAgentSettings

### PR DESCRIPTION
**Pull request by mosche**
Both tracing and metrics require `recording` to be enabled to work. However, if enabling tracing without enabling metrics, `recording` is not enabled. Similarly, if disabling metrics, `recording` may only be disabled if tracing is not enabled.

### CODE REVIEW

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java [hunk 1]
```diff
@@ -61,12 +67,18 @@ public void addClusterSettingsListeners(
     }
 
     /**
-     * Copies APM settings from the provided settings object into the corresponding system properties.
+     * Initialize APM settings from the provided settings object into the corresponding system properties.
+     * Later updates to these settings are synchronized using update consumers.
      * @param settings the settings to apply
      */
-    public void syncAgentSystemProperties(Settings settings) {
-        this.setAgentSetting("recording", Boolean.toString(APM_ENABLED_SETTING.get(settings)));
```
Comment (mosche) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:68]: This looks also problematic, `recording` was previously only enabled if tracing was when initially synching the agent settings.

That makes me wonder if this is needed at all? It looks like the update consumers are triggering at startup which would then typically enable `recording` (assuming metrics are generally enabled).

**Comment by stu-elastic** [comment]
> String key = Strings.split(qualifiedKey, APM_AGENT_SETTING_PREFIX)[1];

This is not going to work when there are two valid prefixes.  The old apm prefix and the new telemetry prefix

**Comment by mosche** [comment]
Yes, when adding the new prefix, we'd have to try splitting on the 2nd one if the 1st split returns `null`.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java [hunk 2]
```diff
@@ -220,22 +232,23 @@ public void setAgentSetting(String key, String value) {
...
-                    }
-                    throw new IllegalArgumentException("Configuration [" + qualifiedKey + "] is either prohibited or unknown.");
+        APM_AGENT_SETTING_PREFIX,
+        (qualifiedKey) -> new Setting<>(qualifiedKey, "", (value) -> {
+            if (qualifiedKey.equals("_na_")) {
+                return value; // shortcut for initialization of delegate
+            }
+            String key = Strings.split(qualifiedKey, APM_AGENT_SETTING_PREFIX)[1];
+            if (PERMITTED_AGENT_KEYS.contains(key) == false) {
+                // TODO figure out why those settings are kept, these should be reformatted / removed by now
```
Comment (pgomulka) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:244]: I wonder what 'by now' in this todo means?
do you think we can clarify this todo?

**Review by pgomulka — APPROVED** [review]
LGTM, left one question but it can be followed up if necessary separately

**Comment by pgomulka** [comment]
> String key = Strings.split(qualifiedKey, APM_AGENT_SETTING_PREFIX)[1];

> This is not going to work when there are two valid prefixes. The old apm prefix and the new telemetry prefix


 this probably we have to wait for https://github.com/elastic/elasticsearch/pull/104345 before merging this pr, we might need to update it. I missed this comment before approving.

**Review by pgomulka — CHANGES_REQUESTED** [review]
String key = Strings.split(qualifiedKey, APM_AGENT_SETTING_PREFIX)[1];

This is not going to work when there are two valid prefixes. The old apm prefix and the new telemetry prefix

this probably we have to wait for https://github.com/elastic/elasticsearch/pull/104345 before merging this pr, we might need to update it. I missed this comment before approving.

**Comment by mosche** [comment]
Yes, I'll wait for #104345 👍

**Comment by mosche** [comment]
I've updated this to only correctly set `recording`. The other part was already fixed in https://github.com/elastic/elasticsearch/pull/104345.

File: modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java [hunk 3]
```diff
@@ -8,81 +8,179 @@
...
+    APMTelemetryProvider apmTelemetryProvider = mock(Mockito.RETURNS_DEEP_STUBS);
 
     /**
      * Check that when the tracer is enabled, it also sets the APM agent's recording system property to true.
      */
     public void testEnableTracing() {
-        APMAgentSettings apmAgentSettings = spy(new APMAgentSettings());
-        Settings settings = Settings.builder().put(APMAgentSettings.TELEMETRY_TRACING_ENABLED_SETTING.getKey(), true).build();
-        apmAgentSettings.syncAgentSystemProperties(settings);
+        boolean metricsEnabled = randomBoolean();
```
Comment (stu-elastic) [inline modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java:54]: No need for randomness.  We can test both cases.

File: modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java [hunk 4]
```diff
@@ -8,81 +8,179 @@
...
-        apmAgentSettings.syncAgentSystemProperties(settings);
+        Settings settings = Settings.builder().put(TRACING_APM_ENABLED_SETTING.getKey(), true).build();
+        apmAgentSettings.initAgentSystemProperties(settings);
 
         verify(apmAgentSettings).setAgentSetting("recording", "true");
+        verify(apmAgentSettings).setAgentSetting("instrument", "true");
+    }
+
+    public void testEnableMetrics() {
+        boolean tracingEnabled = randomBoolean();
```
Comment (stu-elastic) [inline modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java:81]: No need for randomness.  We can test both cases.

File: modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java [hunk 5]
```diff
@@ -8,81 +8,179 @@
...
 
     /**
-     * Check that when the tracer is disabled, it also sets the APM agent's recording system property to false.
+     * Check that when the tracer is disabled, it also sets the APM agent's recording system property to false unless metrics are enabled.
      */
     public void testDisableTracing() {
-        APMAgentSettings apmAgentSettings = spy(new APMAgentSettings());
-        Settings settings = Settings.builder().put(APMAgentSettings.TELEMETRY_TRACING_ENABLED_SETTING.getKey(), false).build();
-        apmAgentSettings.syncAgentSystemProperties(settings);
+        boolean metricsEnabled = randomBoolean();
```
Comment (stu-elastic) [inline modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java:102]: No need for randomness.  We can test both cases.

File: modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java [hunk 6]
```diff
@@ -8,81 +8,179 @@
...
-        apmAgentSettings.syncAgentSystemProperties(settings);
+        Settings settings = Settings.builder().put(TRACING_APM_ENABLED_SETTING.getKey(), false).build();
+        apmAgentSettings.initAgentSystemProperties(settings);
 
         verify(apmAgentSettings).setAgentSetting("recording", "false");
+        verify(apmAgentSettings).setAgentSetting("instrument", "false");
+    }
+
+    public void testDisableMetrics() {
+        boolean tracingEnabled = randomBoolean();
```
Comment (stu-elastic) [inline modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java:129]: No need for randomness.  We can test both cases.

**Review by stu-elastic — APPROVED** [review]
LGTM with randomness removed.

**Review by pgomulka — APPROVED** [review]
LGTM, 
agree with the comments on not using randomisation

**Pull request merged**